src/ZkFold/Symbolic/Data/ByteString.hs

```diff
@@ -100,31 +100,24 @@ class Truncate a b where
     truncate :: a -> b
 
 
--- | Describes types that can increase their capacity by adding zero bits to the beginning (i.e. before the higher register).
---
-class Extend a b where
-    extend :: a -> b
-
-instance (Finite a, ToConstant a Natural, KnownNat n) => ToConstant (ByteString n a) Natural where
-    toConstant (ByteString x xs) = Haskell.foldl (\y p -> toConstant p + base * y) 0 (x:xs)
-        where base = 2 Haskell.^ maxBitsPerRegister @a @n
+instance (ToConstant a Natural) => ToConstant (ByteString n a) Natural where
+    toConstant (ByteString bits) = Haskell.foldl (\y p -> toConstant p + base * y) 0 bits
+        where base = 2
 
 
-instance (FromConstant Natural a, Finite a, KnownNat n) => FromConstant Natural (ByteString n a) where
+instance (FromConstant Natural a, KnownNat n) => FromConstant Natural (ByteString n a) where
 
-    -- | Pack a ByteString as tightly as possible, allocating the largest possible number of bits to each register.
+    -- | Pack a ByteString using one field element per bit.
```

Comment:
Indeed, we lose free conversions to and from UInt, but instead we get much cheaper bitwise operations (no need to run `toBits` before and `fromBits` after the operation) and completely free rotations and resizes. This greatly reduces the number of constraints generated.

Reply:
Would UInt stand any benefit from similar loosening?

Reply:
We will eventually do range checks differently (i.e., with lookup argument instead of bit checks). On the other hand, arithmetic operations can be performed on packed uints.